src/lib.rs

```diff
@@ -0,0 +1,71 @@
+#![cfg_attr(not(test), no_std)]
+
+//! Gimlet Inspector protocol definition.
+//!
+//! The types in this file are intended to be encoded with `hubpack` and stuffed
+//! into a UDP packet. Messages to Gimlet use the `Request` type. Responses from
+//! Gimlet use the `*Response` types -- a response structure may be specific to
+//! the request. In both cases, certain requests/responses may append binary
+//! data _after_ the `hubpack`-encoded data. This is documented below on the
+//! specific items.
+
+use hubpack::SerializedSize;
+use serde::{Serialize, Deserialize};
+
+/// Request format to the inspector agent.
+///
+/// This is an enum so that we implicitly get a protocol version field at the
+/// start of all encoded packets. In the event that multiple versions coexist at
+/// one time, we can add multiple variants. As versions are obsoleted, we can
+/// replace them with stub variants to ensure their indexes don't get reused.
+///
+/// The downside of this approach is that it will fail at 256 versions.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, SerializedSize)]
+pub enum Request {
+    /// A request in v0 consists only of the name of the query to be issued.
+    V0(Query),
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, SerializedSize)]
+pub enum Query {
+    /// Asks the agent to interrogate the sequencer FPGA and send the register
+    /// contents back. The response is always a `SequencerRegistersResponse`.
+    SequencerRegisters,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, SerializedSize)]
+pub enum SequencerRegistersResponse {
```

Comment:
It looks like this response type isn't versioned.  What happens if we add additional variants to this enum?

Reply:
Fair. It's kind of implicitly versioned v0. I could rename the type to make that clear. Adding variants to this enum that are used in the firmware would cause existing clients to become incompatible if the newly modeled error case happens. There's also an ordering sensitivity that I will document.

I hesitate to go "full MGS" on this because this protocol is _intended to be_ a potentially evolving diagnostic protocol used by tools with access to the archive version (Humility, primarily). I actually considered making both requests and responses textual strings for maximum flexibility, but this produced significantly smaller code on the SP side. I'm open to changes in approach if you have insights.